### PR TITLE
Overhaul how chat memory is configured

### PR DESCRIPTION
**Pull request by geoand**
The basic ideas are:

* A default `ChatMemoryStore` bean using in memory storage is now provided 
  - This can be overridden by users simply by providing their own bean
* A default `ChatMemoryProvider` bean is now provided 
  - It uses the `ChatMemoryStore` bean under the hood 
  - It also leverages configuration for use either `MessageWindowChatMemory` or `TokenWindowChatMemory`

`ChatMemory` objects continue to be removed when the AI service class goes out of scope, however now there is also now a programmatic way of removing `ChatMemory` by using `ChatMemoryRemover`. This is necessary for when the AI service is not request scoped.

### CODE REVIEW

File: core/runtime/src/main/java/io/quarkiverse/langchain4j/ChatMemoryRemover.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    private static final Object[] EMPTY_OBJECT_ARRAY = new Object[0];
+
+    private ChatMemoryRemover() {
+    }
+
+    /**
+     * @param aiService The bean that implements the AI Service annotated with {@link RegisterAiService}
+     * @param memoryId The object used as memory IDs for which the corresponding {@link ChatMemory} should be removed
+     */
+    public static void remove(Object aiService, Object memoryId) {
```
Comment (geoand) [inline core/runtime/src/main/java/io/quarkiverse/langchain4j/ChatMemoryRemover.java:24]: I don't like using two objects here as users could easily swap them, but I don't have a better way of handling it...

**Comment by maxandersen** [comment]
context (no pun intended) For  #47 :)

File: docs/modules/ROOT/pages/ai-services.adoc [hunk 2]
```diff
@@ -175,24 +176,40 @@ public class MyChatModelSupplier implements Supplier<ChatLanguageModel> {
...
+The `chatMemoryProviderSupplier` attribute of the `@RegisterAiService` annotation enables configuring the `dev.langchain4j.memory.chat.ChatMemoryProvider`. The default value of this annotation is `RegisterAiService.BeanIfExistsChatMemoryProviderSupplier.class`
 which means that the `AiService` will use whatever `ChatMemoryProvider` bean is configured by the application, while falling back to no memory if no such bean exists.
-An example of such a bean is:
+
+The extension provides a default implementation of `ChatMemoryProvider` which does two things:
+
+* It uses whatever bean `dev.langchain4j.store.memory.chat.ChatMemoryStore` bean is configured, as the backing store. The default implementation is `dev.langchain4j.store.memory.chat.InMemoryChatMemoryStore`
+** If the application provides its own `ChatMemoryStore` bean, that will be used instead of the default `InMemoryChatMemoryStore`,
+* It leverages the available configuration options under `quarkus.langchain4j.chat-memory` to construct the  `ChatMemoryProvider`.
+** The default configuration values result in the usage of `dev.langchain4j.memory.chat.MessageWindowChatMemory` with a window size of ten
```
Comment (maxandersen) [inline docs/modules/ROOT/pages/ai-services.adoc:187]: is there a way to make it unbounded window?

if user wanted to keep the window but instead passivate/persist memory that have not been accessed for a while is that where ChatMemoryRemover come into play?

**Review by cescoffier — APPROVED** [review]

File: docs/modules/ROOT/pages/ai-services.adoc (hunk 2 above)
Comment (geoand) [reply docs/modules/ROOT/pages/ai-services.adoc:187]: > is there a way to make it unbounded window?

Nope, Langchain4j doesn't have that - you would need to write you own

> if user wanted to keep the window but instead passivate/persist memory that have not been accessed for a while is that where ChatMemoryRemover come into play?

Nope, see the updates samples that use `ChatMemoryRemover` to see when it's meant to be used

Comment (geoand) [reply docs/modules/ROOT/pages/ai-services.adoc:187]: When we have clear scopes (or when the existing request scope is used), everything works magically 😄.


I mean that the extension already has the logic to remove chat messages from the store

**Comment by geoand** [comment]
I'd like @andreas-eberle to also have a look since he raised the concern about `ChatMemoryStore` yesterday

**Comment by geoand** [comment]
Once this is in, I'll do a `0.4.0` release

File: docs/modules/ROOT/examples/io/quarkiverse/langchain4j/samples/CustomProvider.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    @Override
+    public ChatMemoryProvider get() {
+        return new ChatMemoryProvider() {
+
+            @Override
+            public ChatMemory get(Object memoryId) {
+                return MessageWindowChatMemory.builder()
+                        .maxMessages(20)
+                        .id(memoryId)
+                        .chatMemoryStore(new InMemoryChatMemoryStore())
```
Comment (andreas-eberle) [inline docs/modules/ROOT/examples/io/quarkiverse/langchain4j/samples/CustomProvider.java:20]: Does it make sense to create a new store here every time?

**Comment by andreas-eberle** [comment]
Looks good to me. Thanks @geoand !

File: docs/modules/ROOT/examples/io/quarkiverse/langchain4j/samples/CustomProvider.java (hunk 3 above)
Comment (geoand) [reply docs/modules/ROOT/examples/io/quarkiverse/langchain4j/samples/CustomProvider.java]: Good catch! I fixed it

**Pull request merged**